File: src/audio-io.ts
import * as fs from 'fs';
import * as path from 'path';

const MIME_BY_EXT: Record<string, string> = {
  '.wav': 'audio/wav',
  '.mp3': 'audio/mpeg',
  '.mp4': 'audio/mp4',
  '.m4a': 'audio/mp4',
  '.aac': 'audio/aac',
  '.flac': 'audio/flac',
  '.ogg': 'audio/ogg',
  '.oga': 'audio/ogg',
  '.opus': 'audio/ogg',
  '.webm': 'audio/webm',
};

export function readStdinBuffer(stdin: NodeJS.ReadableStream): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    const chunks: Buffer[] = [];
    stdin.on('data', (chunk: Buffer | string) => {
      chunks.push(typeof chunk === 'string' ? Buffer.from(chunk) : chunk);
    });
    stdin.on('end', () => resolve(Buffer.concat(chunks)));
    stdin.on('error', reject);
  });
}

export function readStdinText(stdin: NodeJS.ReadableStream): Promise<string> {
  return new Promise((resolve, reject) => {
    let data = '';
    stdin.setEncoding?.('utf-8');
    stdin.on('data', (chunk: string | Buffer) => {
      data += typeof chunk === 'string' ? chunk : chunk.toString('utf-8');
    });
    stdin.on('end', () => resolve(data));
    stdin.on('error', reject);
  });
}

export interface AudioInputOptions {
  filename?: string;
  contentType?: string;
}

export interface AudioInputDeps {
  stdin?: NodeJS.ReadableStream;
  readFile?: (path: string) => Buffer;
  fileExists?: (path: string) => boolean;
}

export interface AudioInput {
  data: Buffer;
  filename: string;
  contentType: string;
}

export async function resolveAudioInput(
  fileArg: string | undefined,
  opts: AudioInputOptions,
  deps: AudioInputDeps,
): Promise<AudioInput> {
  const readFile = deps.readFile ?? ((p: string) => fs.readFileSync(p));
  const fileExists = deps.fileExists ?? ((p: string) => fs.existsSync(p));
  const stdin = deps.stdin ?? process.stdin;

  if (fileArg && fileArg !== '-') {
    if (!fileExists(fileArg)) {
      throw new Error(`Audio file not found: ${fileArg}`);
    }
    const ext = path.extname(fileArg).toLowerCase();
    return {
      data: readFile(fileArg),
      filename: opts.filename ?? path.basename(fileArg),
      contentType:
        opts.contentType ?? MIME_BY_EXT[ext] ?? 'application/octet-stream',
    };
  }

  const data = await readStdinBuffer(stdin);
  if (data.length === 0) {
    throw new Error(
      'No audio input. Pass a file path or pipe audio to stdin (`--filename` recommended when piping).',
    );
  }
  const filename = opts.filename ?? 'audio.wav';
  const ext = path.extname(filename).toLowerCase();
  return {
    data,
    filename,
    contentType:
      opts.contentType ?? MIME_BY_EXT[ext] ?? 'application/octet-stream',
  };
}

interface HttpResponseLike {
  ok: boolean;
  status: number;
  statusText: string;
  text: () => Promise<string>;
}

export async function writeApiErrorAndExit(
  response: HttpResponseLike,
  stderr: NodeJS.WritableStream,
): Promise<never> {
  const body = await response.text();
  stderr.write(`API error ${response.status} ${response.statusText}\n`);
  stderr.write(body);
  if (!body.endsWith('\n')) stderr.write('\n');
  process.exit(1);
}

export interface BinaryOutputDeps {
  stdout?: NodeJS.WritableStream;
  stderr?: NodeJS.WritableStream;
  writeFile?: (path: string, data: Buffer) => void;
  stdoutIsTTY?: boolean;
}

export function ensureNotWritingBinaryToTTY(
  output: string | undefined,
  deps: BinaryOutputDeps,
): void {
  const isTTY = deps.stdoutIsTTY ?? Boolean((process.stdout as any).isTTY);
  if (!output && isTTY) {
    throw new Error(
      'Refusing to write binary audio to a terminal. Use `-o <path>` or pipe stdout to a file.',
    );
  }
}

export function writeBinaryOutput(
  buffer: Buffer,
  output: string | undefined,
  deps: BinaryOutputDeps,
): void {
  const stdout = deps.stdout ?? process.stdout;
  const stderr = deps.stderr ?? process.stderr;
  const writeFile = deps.writeFile ?? ((p: string, d: Buffer) => fs.writeFileSync(p, d));

  if (output) {
    writeFile(output, buffer);
    stderr.write(`Wrote ${buffer.length} bytes to ${output}\n`);
  } else {
    stdout.write(buffer);
  }
}

File: src/audio_io.rs
//! Shared audio I/O: input resolution (file or stdin), MIME guessing, and
//! binary output handling with a TTY guard.

use std::io::{IsTerminal, Read, Write};
use std::path::Path;

use crate::error::{AppError, Result};
use crate::net::Resp;

fn mime_by_ext(ext: &str) -> Option<&'static str> {
    match ext {
        ".wav" => Some("audio/wav"),
        ".mp3" => Some("audio/mpeg"),
        ".mp4" => Some("audio/mp4"),
        ".m4a" => Some("audio/mp4"),
        ".aac" => Some("audio/aac"),
        ".flac" => Some("audio/flac"),
        ".ogg" => Some("audio/ogg"),
        ".oga" => Some("audio/ogg"),
        ".opus" => Some("audio/ogg"),
        ".webm" => Some("audio/webm"),
        _ => None,
    }
}

fn dotted_ext(name: &str) -> String {
    match Path::new(name).extension().and_then(|e| e.to_str()) {
        Some(e) => format!(".{}", e.to_lowercase()),
        None => String::new(),
    }
}

pub struct AudioInput {
    pub data: Vec<u8>,
    pub filename: String,
    pub content_type: String,
}

pub fn read_stdin_buffer() -> Result<Vec<u8>> {
    let mut buf = Vec::new();
    std::io::stdin().read_to_end(&mut buf)?;
    Ok(buf)
}

pub fn read_stdin_text() -> Result<String> {
    let mut s = String::new();
    std::io::stdin().read_to_string(&mut s)?;
    Ok(s)
}

pub fn resolve_audio_input(
    file_arg: Option<&str>,
    filename_opt: Option<&str>,
    content_type_opt: Option<&str>,
) -> Result<AudioInput> {
    if let Some(file) = file_arg {
        if file != "-" {
            let path = Path::new(file);
            if !path.exists() {
                return Err(AppError::new(format!("Audio file not found: {file}")));
            }
            let ext = dotted_ext(file);
            let data = std::fs::read(path)?;
            let filename = filename_opt.map(str::to_string).unwrap_or_else(|| {
                path.file_name()
                    .and_then(|n| n.to_str())
                    .unwrap_or("")
                    .to_string()
            });
            let content_type = content_type_opt
                .map(str::to_string)
                .or_else(|| mime_by_ext(&ext).map(str::to_string))
                .unwrap_or_else(|| "application/octet-stream".to_string());
            return Ok(AudioInput {
                data,
                filename,
                content_type,
            });
        }
    }

    let data = read_stdin_buffer()?;
    if data.is_empty() {
        return Err(AppError::new(
            "No audio input. Pass a file path or pipe audio to stdin (`--filename` recommended when piping).",
        ));
    }
    let filename = filename_opt.unwrap_or("audio.wav").to_string();
    let ext = dotted_ext(&filename);
    let content_type = content_type_opt
        .map(str::to_string)
        .or_else(|| mime_by_ext(&ext).map(str::to_string))
        .unwrap_or_else(|| "application/octet-stream".to_string());
    Ok(AudioInput {
        data,
        filename,
        content_type,
    })
}

/// Print an API error (status + body) to stderr and exit 1.
pub fn write_api_error_and_exit(resp: &Resp) -> ! {
    let body = resp.text();
    eprintln!("API error {} {}", resp.status, resp.status_text);
    eprint!("{body}");
    if !body.ends_with('\n') {
        eprintln!();
    }
    std::process::exit(1);
}

pub fn ensure_not_writing_binary_to_tty(output: Option<&str>) -> Result<()> {
    if output.is_none() && std::io::stdout().is_terminal() {
        return Err(AppError::new(
            "Refusing to write binary audio to a terminal. Use `-o <path>` or pipe stdout to a file.",
        ));
    }
    Ok(())
}

pub fn write_binary_output(buffer: &[u8], output: Option<&str>) -> Result<()> {
    match output {
        Some(path) => {
            std::fs::write(path, buffer)?;
            eprintln!("Wrote {} bytes to {}", buffer.len(), path);
        }
        None => {
            let mut stdout = std::io::stdout();
            stdout.write_all(buffer)?;
            stdout.flush()?;
        }
    }
    Ok(())
}
